protocol/contracts/templegold/DaiGoldAuction.sol

```diff
@@ -0,0 +1,339 @@
+            revert CannotStartAuction();
+        }
+        _distributeGold();
+        uint256 totalGoldAmount = nextAuctionGoldAmount;
+        nextAuctionGoldAmount = 0;
+        uint256 epochId = _currentEpochId = _currentEpochId + 1;
+        
+        if (totalGoldAmount < config.auctionMinimumDistributedGold) { revert LowGoldDistributed(totalGoldAmount); }
+
+        EpochInfo storage info = epochs[epochId];
+        info.totalAuctionTokenAmount = totalGoldAmount;
+        uint128 startTime = info.startTime = uint128(block.timestamp) + config.auctionStartCooldown;
+        uint128 endTime = info.endTime = startTime + AUCTION_DURATION;
+
+        emit AuctionStarted(epochId, msg.sender, startTime, endTime, totalGoldAmount);
+    }
+
+    /**
+     * @notice Deposit bidding token for current running epoch auction
+     * @param amount Amount of bid token to deposit
+     */
+    function bid(uint256 amount) external virtual override onlyWhenLive {
+        if (amount == 0) { revert CommonEventsAndErrors.ExpectedNonZero(); }
+
+        bidToken.safeTransferFrom(msg.sender, treasury, amount);
+
+        uint256 epochIdCache = _currentEpochId;
+        depositors[msg.sender][epochIdCache] += amount;
+
+        EpochInfo storage info = epochs[epochIdCache];
+        info.totalBidTokenAmount += amount;
+        emit Deposit(msg.sender, epochIdCache, amount);
+    }
+
+    /**
+     * @notice Claim share of Temple Gold for epoch
+     * Can only claim for past epochs, not current auction epoch.
+     * @param epochId Id of epoch
+     */
+    function claim(uint256 epochId) external virtual override {
+        /// @notice cannot claim for current live epoch
+        EpochInfo storage info = epochs[epochId];
+        if (!info.hasEnded()) { revert CannotClaim(epochId); }
+        /// @dev epochId could be invalid. eg epochId > _currentEpochId
+        if (info.startTime == 0) { revert InvalidEpoch(); }
+        if (claimed[msg.sender][epochId]) { revert AlreadyClaimed(); }
+
+        uint256 bidTokenAmount = depositors[msg.sender][epochId];
+        if (bidTokenAmount == 0) { revert CommonEventsAndErrors.ExpectedNonZero(); }
+        claimed[msg.sender][epochId] = true;
+        uint256 claimAmount = bidTokenAmount.mulDivRound(info.totalAuctionTokenAmount, info.totalBidTokenAmount, false);
+        claimedAmount[msg.sender][epochId] = claimAmount;
+        templeGold.safeTransfer(msg.sender, claimAmount);
+        emit Claim(msg.sender, epochId, bidTokenAmount, claimAmount);
+    }
+
+    /**
+     * @notice Checkpoint total Temple Gold distributed for next epoch.
+     * Can only be called by Temple Gold contract or elevated access (for rebalancing purpose)
+     * @param amount Amount of Temple Gold to checkpoint
+     */
+    function notifyDistribution(uint256 amount) external override {
+        if (msg.sender != address(templeGold) && !isElevatedAccess(msg.sender, msg.sig)) { revert CommonEventsAndErrors.InvalidAccess(); }
```

Comment:
I don't love the fact that ElevatedAccess can call `notifyDistribution()` with an `amount` without actually sending that amount, and make this contract insolvent. 

But I understand this gives you also the flexibility to "notify" of donated templeGold tokens so that you don't have to complicate the recoverToken function even more. 

(But I do want to point out that misusing this function can break the contract's solvency so that ElevatedAccess is extremely aware, hehe)